agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+// which means it must never take that lock itself or call anything that does.
+func (c *CAManager) getIntermediateCASigned(provider ca.Provider, newActiveRoot *structs.CARoot) error {
+	csr, err := provider.GenerateIntermediateCSR()
+	if err != nil {
+		return err
+	}
+
+	var intermediatePEM string
+	if err := c.srv.forwardDC("ConnectCA.SignIntermediate", c.srv.config.PrimaryDatacenter, c.srv.generateCASignRequest(csr), &intermediatePEM); err != nil {
+		// this is a failure in the primary and shouldn't be capable of erroring out our establishing leadership
+		c.logger.Warn("Primary datacenter refused to sign our intermediate CA certificate", "error", err)
+		return nil
+	}
+
+	if err := provider.SetIntermediate(intermediatePEM, newActiveRoot.RootCert); err != nil {
+		return fmt.Errorf("Failed to set the intermediate certificate with the CA provider: %v", err)
+	}
+
+	intermediateCert, err := connect.ParseCert(intermediatePEM)
+	if err != nil {
+		return fmt.Errorf("error parsing intermediate cert: %v", err)
+	}
+
+	// Append the new intermediate to our local active root entry. This is
+	// where the root representations start to diverge.
+	newActiveRoot.IntermediateCerts = append(newActiveRoot.IntermediateCerts, intermediatePEM)
+	newActiveRoot.SigningKeyID = connect.EncodeSigningKeyID(intermediateCert.SubjectKeyId)
+
+	c.logger.Info("received new intermediate certificate from primary datacenter")
+	return nil
+}
+
+// intermediateCertRenewalWatch periodically attempts to renew the intermediate cert.
+func (c *CAManager) intermediateCertRenewalWatch(ctx context.Context) error {
+	isPrimary := c.srv.config.Datacenter == c.srv.config.PrimaryDatacenter
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(structs.IntermediateCertRenewInterval):
+			retryLoopBackoffAbortOnSuccess(ctx, func() error {
+				return c.RenewIntermediate(isPrimary)
+			}, func(err error) {
+				c.logger.Error("error renewing intermediate certs",
+					"routine", intermediateCertRenewWatchRoutineName,
+					"error", err,
+				)
+			})
+		}
+	}
+}
+
+// RenewIntermediate checks the intermediate cert for
+// expiration. If more than half the time a cert is valid has passed,
+// it will try to renew it.
+func (c *CAManager) RenewIntermediate(isPrimary bool) error {
+	// Grab the 'lock' right away so the provider/config can't be changed out while we check
+	// the intermediate.
+	if err := c.setState(CAStateRenewIntermediate, true); err != nil {
+		return err
+	}
+	defer c.setState(CAStateReady, false)
+
+	provider, _ := c.getCAProvider()
+	if provider == nil {
+		// this happens when leadership is being revoked and this go routine will be stopped
+		return nil
+	}
+	// If this isn't the primary, make sure the CA has been initialized.
+	if !isPrimary && !c.configuredSecondaryCA() {
+		return fmt.Errorf("secondary CA is not yet configured.")
+	}
+
+	state := c.srv.fsm.State()
+	_, activeRoot, err := state.CARootActive(nil)
```

Comment:
we should deep-copy `activeRoot` here as `newActiveRoot` and use that to mutate. I think we're currently mutating the memdb copy which is usually a bad sign